sklearn/inspection/tests/test_partial_dependence.py

```diff
@@ -836,3 +836,55 @@ def test_kind_average_and_average_of_individual(Estimator, data):
     pdp_ind = partial_dependence(est, X=X, features=[1, 2], kind="individual")
     avg_ind = np.mean(pdp_ind["individual"], axis=1)
     assert_allclose(avg_ind, pdp_avg["average"])
+
+
+def test_grid_from_X_missing_values_numerical_features():
+    """Check the behavior when X contains missing values with numerical features.
+
+    When generating the grid, we should ignore the missing values to avoid introducing
+    a bias in the partial dependence computation. Therefore, we make sure that there is
+    no missing values in the grid.
+
+    Non-regression test for:
+    https://github.com/scikit-learn/scikit-learn/issues/25401
+    """
+    rng = np.random.RandomState(0)
+    percentiles, grid_resolution, is_categorical = (0.05, 0.95), 5, [False, False]
+    X = rng.randn(100, 2)
+    X[::5] = np.nan
+
+    grid, axes = _grid_from_X(X, percentiles, is_categorical, grid_resolution)
+    assert not all([is_scalar_nan(elt) for elt in grid.ravel()])
+    for axis in axes:
+        assert not all([is_scalar_nan(elt) for elt in axis])
```

Comment:
I was surprised to see `not all(..)`, I was expecting a `not any(...)`. The expectation is based on the fact that the code that is being tested seems to filter out all NaNs with a `e for e in values if not is_scalar_nan(e)` and I can't get `nanquantile` to return a NaN as its return value. Did I misunderstand what the code does or is there a scenario where we could have a NaN here?